gift-certificate-web/src/main/java/com/epam/esm/rest/controller/GiftCertificateController.java

```diff
@@ -0,0 +1,137 @@
+package com.epam.esm.rest.controller;
+
+import com.epam.esm.entity.Tag;
+import com.epam.esm.model.SearchParametersHolder;
+import com.epam.esm.model.TaggedGiftCertificate;
+import com.epam.esm.rest.exceptions.ResourceNotFoundException;
+import com.epam.esm.services.GiftCertificateService;
+import com.epam.esm.services.exceptions.ServiceException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@RestController
+@RequestMapping("/api")
+public class GiftCertificateController {
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    private GiftCertificateService giftCertificateService;
+
+    @Autowired
+    public GiftCertificateController(GiftCertificateService giftCertificateService) {
+        this.giftCertificateService = giftCertificateService;
+    }
+
+    @PostMapping("/certificates")
+    public TaggedGiftCertificate savePortfolio(@RequestBody TaggedGiftCertificate giftCertificate) {
+        TaggedGiftCertificate returnObject;
+        try {
+            Optional<TaggedGiftCertificate> optionalGiftCertificate = giftCertificateService.save(giftCertificate);
+            returnObject = optionalGiftCertificate.orElseThrow(() -> new RuntimeException("Operation failed."));
+        } catch (ServiceException e) {
+            LOGGER.error("save error: " + e.getMessage());
```

Comment:
In some cases, you have triple logging. So, try to move the whole log writing in one place or layout.